src/ws/ws-client.ts

```diff
@@ -0,0 +1,251 @@
+/**
+ * Graasp websocket client
+ * Provides front-end integration for real-time updates using WebSocket
+ * Implements the client protocol from https://github.com/graasp/graasp-websockets
+ * 
+ * @author Alexandre CHAU
+ */
+
+import { ClientMessage, ServerMessage } from "graasp-websockets/src/interfaces/message";
+import { QueryClientConfig } from "../types";
+
+export type Channel = {
+    entity: "item" | "member";
+    name: string,
+};
+type UpdateHandlerFn = (data: ServerMessage) => void;
+
+/**
+ * Helper to remove the first element in an array that
+ * matches the provided value IN PLACE
+ */
+function arrayRemoveFirstEqual<T>(array: Array<T>, value: T, eqFn = (a: T, b: T) => a === b) {
+    const pos = array.findIndex(v => eqFn(v, value));
+    const removed = array.splice(pos, 1);
+    return (removed.length > 0);
+}
+
+/**
+ * Helper to add a value to a mapped array, creating the array
+ * if the map entry does not exist
+ */
+function addToMappedArray<S, T>(map: Map<S, Array<T>>, key: S, value: T) {
+    const array = map.get(key);
+    if (array === undefined) {
+        map.set(key, [value]);
+    } else {
+        array.push(value);
+    }
+}
+
+/**
+ * Helper to convert a channel object into a unique string
+ * (deep equality) to serve as map keys
+ */
+function buildChannelKey(channel: Channel): string {
+    return JSON.stringify(channel);
```

Comment:
This is actually not that great. If the `channel` parameter (which can be provided by many different producers such as server response, ws-client consumers, ...) is not an object which is exactly identical (including order of keys!) then the map key equality check will fail, see example below:
![image](https://user-images.githubusercontent.com/14943421/123286618-59ee9b00-d50e-11eb-97bc-bd2f1f7a5058.png)

Instead, create custom serializable string that will always be the same (make sure that there is no conflicting semantics with e.g. channel names or entity types)

Reply:
Fixed in 0dc7684